packages/order-utils/src/order_validation_utils.ts

```diff
@@ -171,6 +179,81 @@ export class OrderValidationUtils {
         const provider = providerUtils.standardizeOrThrow(supportedProvider);
         this._provider = provider;
     }
+
+    // TODO(xianny): remove this method once the smart contracts have been refactored
+    // to return helpful revert reasons instead of ORDER_UNFILLABLE. Instruct devs
+    // to make "calls" to validate order fillability + getOrderInfo for fillable amount.
+    // This method recreates functionality from ExchangeWrapper (@0x/contract-wrappers < 11.0.0)
+    // to make migrating easier in the interim.
+    /**
+     * Validate if the supplied order is fillable, and throw if it isn't
+     * @param erc20ProxyAddress A valid address where an ERC20ProxyContract is deployed
+     * @param erc721ProxyAddress A valid address where an ERC721ProxyContract is deployed
+     * @param exchangeAddress A valid address where the ExchangeContract is deployed
+     * @param zrxTokenAddress A valid address where ZRXTokenContract is deployed
+     * @param networkId The network for all contracts
+     * @param provider The same provider used to interact with contracts
+     * @param signedOrder SignedOrder of interest
+     * @param opts ValidateOrderFillableOpts options (e.g expectedFillTakerTokenAmount.
+     * If it isn't supplied, we check if the order is fillable for the remaining amount.
+     * To check if the order is fillable for a non-zero amount, set `validateRemainingOrderAmountIsFillable` to false.)
+     */
+    public async simpleValidateOrderFillableOrThrowAsync(
+        erc20ProxyAddress: string,
+        erc721ProxyAddress: string,
+        exchangeAddress: string,
+        zrxTokenAddress: string,
+        networkId: number,
+        provider: SupportedProvider,
+        signedOrder: SignedOrder,
+        opts: ValidateOrderFillableOpts = {},
+    ): Promise<void> {
+        assert.doesConformToSchema('opts', opts, validateOrderFillableOptsSchema);
+
+        const exchange = new ExchangeContract(exchangeAddress, provider);
+        const balanceAllowanceFetcher = new AssetBalanceAndProxyAllowanceFetcher(
+            new ERC20ProxyContract(erc20ProxyAddress, provider),
+            new ERC721ProxyContract(erc721ProxyAddress, provider),
+            provider,
+            BlockParamLiteral.Latest,
+        );
+        const balanceAllowanceStore = new BalanceAndProxyAllowanceLazyStore(balanceAllowanceFetcher);
```

Comment:
`ExchangeTransferSimulator` is useful for basically one reason, If the user is trading XYZ for ZRX and the order has fees, then the `ExchangeTransferSimulator` can handle this case (as it simulates the transfer of ZRX to maker prior to fee payment). This is an annoying case to handle for and is often overlooked, resulting in invalid orders for completely valid orders if only Balances and Allowances are validated. 

I do not believe DevUtils currently handles this edge case.

Extrapolating to V3. I'm selling a Kitty for 100 WETH, there is a 1 WETH fee. I have 0 WETH. This order is valid as there will be funds for fees during settlement but not prior to settlement.

Reply:
@dekz aren't fees paid in ETH _not_ WETH in V3?